challenge_find_my_mates/src/find_my_mates.py

```diff
@@ -0,0 +1,338 @@
+#!/usr/bin/env python
+import rospy
+import math
+import PyKDL as kdl
+from robot_smach_states.util import startup
+from smach import StateMachine, State
+from robot_smach_states import NavigateToWaypoint, SetInitialPose, Initialize, WaitTime, StartChallengeRobust, Say
+from robot_smach_states.util.designators import EntityByIdDesignator
+from robot_skills.util import kdl_conversions
+from hmi import TimeoutException
+from collections import Counter
+from robocup_knowledge import load_knowledge
+
+challenge_knowledge = load_knowledge('challenge_find_my_mates')
+
+STARTING_POINT = challenge_knowledge.starting_point
+ROOM_ID = challenge_knowledge.room
+SEARCH_POINT = challenge_knowledge.search_point
+
+
+class FindPeople(State):
+    """
+    Locate three people in the provided room
+    """
+
+    def __init__(self, robot, search_timeout=60, look_distance=2.0, min_dist=0.5):
+        """ Initialization method
+        :param robot: robot api object
+        :param search_timeout: (float) maximum time the robot is allowed to search
+        :param look_distance: (float) robot only considers laser entities within this radius
+        """
+        State.__init__(self, outcomes=['Done', 'Aborted', 'Failed'])
+
+        self._robot = robot
+        self._search_timeout = search_timeout
+        self._look_distance = look_distance
+        self._min_dist = min_dist
+        self._people = []
+
+    def execute(self, userdata=None):
+        rospy.loginfo("Trying to find my mates")
+        self._robot.head.look_at_standing_person()
+        self._robot.speech.speak("Please look at me while I am looking for you",
+                                 block=False)
+
+        start_time = rospy.Time.now()
+
+        look_distance = 2.0  # magic number 4
+        look_angles = [f * math.pi / d if d != 0 else 0.0 for f in [-1, 1] for d in [0, 6, 4, 2.3]]  # Magic numbers
+        head_goals = [kdl_conversions.VectorStamped(x=look_distance * math.cos(angle),
+                                                    y=look_distance * math.sin(angle), z=1.7,
+                                                    frame_id="/%s/base_link" % self._robot.robot_name)
+                      for angle in look_angles]
+        head_goals.extend([kdl_conversions.VectorStamped(x=2.0 * look_distance * math.cos(angle),
+                                                         y=2.0 * look_distance * math.sin(angle), z=1.7,
+                                                         frame_id="/%s/base_link" % self._robot.robot_name)
+                           for angle in look_angles])
+        n = 0
```

Comment:
What's with the `n` and `i`? Can you explain what each does?

I think for `i` cyou can just enumerate the head goals: 

```python
for i, head_goal in enumerate(head_goals):
   # do stuff
```

Reply:
The ```i``` is copied from ```learn_operator``` (just as most of that smach state), and you want to reset ```i``` to 0 to keep looking until the timeout is reached, which is not possible using enumerate, right? 

And the ```n``` represents the number of people detected, since you want to keep looking until you've found four people. I'll describe that in a comment there.